nftables_test.go

```diff
@@ -2457,6 +2457,151 @@ func TestCreateUseAnonymousSet(t *testing.T) {
 	}
 }
 
+func TestCappedErrMsgOnObj(t *testing.T) {
+	c, newNS := openSystemNFTConn(t)
+	c, err := nftables.New(nftables.WithNetNSFd(int(newNS)), nftables.AsLasting())
+	if err != nil {
+		t.Fatalf("nftables.New() failed: %v", err)
+	}
+	defer cleanupSystemNFTConn(t, newNS)
+	c.FlushRuleset()
+	defer c.FlushRuleset()
+
+	filter := c.AddTable(&nftables.Table{
+		Family: nftables.TableFamilyIPv4,
+		Name:   "filter",
+	})
+	if err := c.Flush(); err != nil {
+		t.Errorf("failed adding table: %v", err)
+	}
+
+	const counterLen = 500
+	counters := make([]nftables.Obj, counterLen)
+	for i := 0; i < counterLen; i++ {
+		counters[i] = &nftables.CounterObj{
+			Table:   filter,
+			Name:    fmt.Sprintf("ctr%d", i),
+			Bytes:   uint64(-1 * i),
+			Packets: uint64(-1 * i),
+		}
+		c.AddObj(counters[i])
+	}
+
+	if err := c.Flush(); err != nil {
+		// this test leverages the fact that read buffer of the netlink socket
+		// is too small to receive all ack messages sent by the kernel
+		//
+		// the Flush method tries to receive all acks after sending the messages
+		// https://github.com/google/nftables/blob/cbeb0fb1eccf9ef582c20982c72e73107d1898a5/conn.go#L185-L193
+		//
+		// buffer peeking implemented by the netlink go library fails with a "no buffer space available"
+		// https://github.com/mdlayher/netlink/blob/7fa043dcb6f27ed7e084dc90ddf5b6d4092478a9/conn_linux.go#L123-L127
+		// this results in part of "non-received" acknowledgments to end up being received
+		// in the following c.GetObj(counters[0]) call
+		//
```

Comment:
Thanks for the additional description and details.

This makes me wonder: could we instead fix this in github.com/mdlayher/netlink by handling the “no buffer space” error and enlarging the buffer, as that code already tries to do? Would that make this entire PR unnecessary?

Reply:
Good thinking but unfortunately no. Even if we would propose a PR fix that would enlarge the buffer, we would still have the issue of the "no error" messages being sent in GetSetElements as initially reported in issue #175 (and including other functions using conn.Receive without retries). Note that the "no error" messages seen in the issue are not caused by the "no buffer space" error. The "no error" messages (basically acks) are a result of using the nftables lasting connection feature. At which point and why we receive these acks in the lasting connection is [still unknown to me](https://github.com/google/nftables/issues/175#issuecomment-1233191956).

What I am doing with this test is only mimicking this behaviour with GetObj function since [the comment reporting the issue with GetObj](https://github.com/google/nftables/issues/175#issuecomment-1252691215) was not specific on how to reproduce the issue so I was not able to write that specific test other than reproducing the "no buffer space" error and forcing acknowledgements as a response for the GetObj function. I know that it seems a bit hacky, but it still shows as a good test for the retry logic.

Let me also note that "no buffer space" error was already discussed once in issue #103 but the issue was closed as stale.

Reply:
> Good thinking but unfortunately no. Even if we would propose a PR fix that would enlarge the buffer, we would still have the issue of the "no error" messages being sent in GetSetElements as initially reported in issue #175 (and including other functions using conn.Receive without retries). Note that the "no error" messages seen in the issue are not caused by the "no buffer space" error. The "no error" messages (basically acks) are a result of using the nftables lasting connection feature. 

I think these two issues might best be fixed separately — the “no buffer space” one in mdlayher/netlink, and the acknowledgement issue in google/nftables.

> At which point and why we receive these acks in the lasting connection is [still unknown to me](https://github.com/google/nftables/issues/175#issuecomment-1233191956).

I think it would be best if we actually understood what the issue with the acks is. It still sounds to me like we’re requesting more acks than we should, or something along these lines.

While your PR probably works, it seems to me like it’s papering over the issue.

Reply:
Agreed, thanks for the input. I have introduced a different approach in PR #194. I will try to address the "no buffer" error separately in netlink library.